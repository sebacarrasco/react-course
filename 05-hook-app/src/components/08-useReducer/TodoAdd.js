import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ( { handleAddTodo } ) => {

    const [ { description }, handleInputChange, reset ] = useForm({
        description: ""
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        // Validación del input
        if ( description.trim().length <= 1)
        {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        // Se le manda el dispatch con la acción y él ya sabe qué
        // reducer usar y hará render de lo que cambie luego de actualizar el state
        handleAddTodo(newTodo)
        reset();
    };

    return (
        <>
            <h4> Add TODO </h4>
            <hr/>

            <form onSubmit={ handleSubmit }>

                <input
                className="form-control"
                type="text"
                name="description"
                placeholder="Do something..."
                autoComplete="off"
                onChange={ handleInputChange }
                value={ description }
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 w-100"
                >
                    Add
                </button>

            </form>
        </>
    )
}
