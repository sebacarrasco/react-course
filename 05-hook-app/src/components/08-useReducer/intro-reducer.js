

const initialState = [{
    id: 1,
    todo: "primera tarea",
    done: false
}];

const todoReducer = (state=initialState, action) => {

    // El  action?.type quiere decir que si action tiene algún valor,
    // entonces le revisa la propiedad type, en caso contrario no se hace nada
    if ( action?.type === "agregar")
    {
        return [...state, action.payload];
    }

    return state;
};

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: "segunda tarea",
    done: false
}

const agregarTodoAction = {
    type: "agregar",
    payload: newTodo // Es solo convención los nombres de type y payload
}

todos = todoReducer(todos, agregarTodoAction);

console.log(todos)