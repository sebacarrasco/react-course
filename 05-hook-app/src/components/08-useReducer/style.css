body{
    padding: 5%
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
}

p{
    cursor: pointer;
    margin: 0px;
}

.complete {
    text-decoration: line-through;
}