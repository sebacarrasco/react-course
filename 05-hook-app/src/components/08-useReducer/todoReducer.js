
export const todoReducer = ( state=[], action ) => {

    switch (action.type) {
        case "add":
            // Acá se manda el objeto
            return [...state, action.payload];
    
        case "delete":
            // acá se manda el id
            return state.filter(task => task.id !== action.payload);

        case "toggle":
            // acá se manda el id

            // Versión larga:
            // return state.map(task => {
            //     if (task.id === action.payload)
            //     {
            //         return {
            //             ...task,
            //             done: !task.done
            //         };
            //     }
            //     else
            //     {
            //         return task;
            //     }
            // });

            // Versión corta
            return state.map( task => 
                ( task.id === action.payload)
                    ?  {...task, done: !task.done }
                    : task
            )
        default:
            return state;
    }

};