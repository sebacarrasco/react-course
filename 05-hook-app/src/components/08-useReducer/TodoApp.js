import React, { useReducer, useEffect } from 'react';
// import { useForm } from '../../hooks/useForm';
import "./style.css";
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
    return JSON.parse(localStorage.getItem("todos")) || [];
}

export const TodoApp = () => {

    // useReducer es una alternativa al useState que se suele usar cuando hay mucha
    // lógica detrás de la modificación del state.
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos))
    }, [todos]);

    const handleDelete = (todoId) => {
        const action = {
            type: "delete",
            payload: todoId
        }

        dispatch(action);
    };

    const handleToggle = (todoId) => {
        dispatch({
            type: "toggle",
            payload: todoId
        });
    };

    const handleAddTodo = ( newTodo ) => {
        dispatch({
            type: "add",
            payload: newTodo
        })
    }

    return (
        <div>
            <h1>TODO App { todos.length } </h1>
            <hr/>

            <div className="row">

                <div className="col-7">
                    <TodoList todos={todos} handleDelete={ handleDelete } handleToggle={ handleToggle }/>
                </div>

            <div className="col-5">
                <TodoAdd handleAddTodo={ handleAddTodo }/>
            </div>

            </div>
        </div>
    )
}
