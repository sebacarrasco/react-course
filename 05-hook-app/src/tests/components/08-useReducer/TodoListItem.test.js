import React from 'react';
import { demoTodos } from '../../fixtures/demoTodos';
import { shallow } from "enzyme";
import { TodoListItem } from '../../../components/08-useReducer/TodoListItem';

describe('Tests in <TodoListItem />', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();
    const todo = demoTodos[0];
    const index = 0;

    let wrapper = shallow(<TodoListItem
        todo={ todo }
        index={ index }
        handleDelete={ () => {} }
        handleToggle={ () => {} }
    />);
    beforeEach( () => {
        jest.clearAllMocks();
        wrapper = shallow(<TodoListItem
            todo={ todo }
            index={ index }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />);
    });
    

    test('should match the snapshot', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('should call the handleDelete function', () => {

        wrapper.find("button").simulate("click");
        expect(handleDelete).toHaveBeenCalledWith(todo.id);

    });

    test('should call the handleToggle function', () => {

        wrapper.find("p").simulate("click");
        expect(handleToggle).toHaveBeenCalledWith(todo.id);
        
    });

    test('should show the text correctly', () => {

        const parragraphText = wrapper.find("p").text().trim();
        expect(parragraphText).toBe(`${ index + 1 }. ${ todo.desc }`)
        
    });

    test('should have the complete class if todo.done is true', () => {
       
        todo.done = true;
        const wrapper = shallow(<TodoListItem
            todo={ todo }
            index={ index }
            handleDelete={ handleDelete }
            handleToggle={ handleToggle }
        />);

        // const {className} = wrapper.find("p").props();
        // expect(className).toContain("complete");
        // Lo comentado es equivalente a:
        expect(wrapper.find("p").hasClass("complete")).toBe(true);
        
    });
    
    
    
    
});
