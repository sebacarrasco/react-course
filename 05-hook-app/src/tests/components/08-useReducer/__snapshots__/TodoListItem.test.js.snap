// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <TodoListItem /> should match the snapshot 1`] = `
<li
  className="list-group-item"
  key="1"
>
  <p
    className="false"
    onClick={[Function]}
  >
    1
    . 
    Learn react
  </p>
  <button
    className="btn btn-danger"
    onClick={[Function]}
  >
    Delete
  </button>
</li>
`;
